docs/Encryption.md

```diff
@@ -0,0 +1,76 @@
+S3Proxy
+
+# Encryption 
+
+## Motivation 
+The motivation behind this implementation is to provide a fully transparent and secure encryption to the s3 client while having the ability to write into different clouds.
+
+## Cipher mode
+The chosen cipher is ```AES/CFB/NoPadding``` because it provides the ability to read from an offset like in the middle of a ```Blob```.
+While reading from an offset the decryption process needs to consider the previous 16 bytes of the AES block.
+
+### Key generation
+The encryption uses a 128-bit key that will be derived from a given password and salt in combination with random initialization vector that will be stored in each part padding.
+
+## How a blob is encrypted 
+Every uploaded part get a padding of 64 bytes that includes the necessary information for decryption. The input stream from a s3 client is passed through ```CipherInputStream``` and piped to append the 64 byte part padding at the end the encrypted stream. The encrypted input stream is then processed by the ```BlobStore``` to save the ```Blob```.
+
+| Name      | Byte size | Description                                                    |
+|-----------|-----------|----------------------------------------------------------------|
+| Delimiter | 8 byte    | The delimiter is used to detect if the ```Blob``` is encrypted |
+| IV        | 16 byte   | AES initialization vector                                      |
+| Part      | 4 byte    | The part number                                                |
+| Size      | 8 byte    | The unencrypted size of the ```Blob```                         |
+| Version   | 2 byte    | Version can be used in the future if changes are necessary     |
+| Reserved  | 26 byte   | Reserved for future use                                        |
```

Comment:
This will not work if you have for example 1000 parts or more as it will occupy the metadata space which is 2kb aur 4kb if I am not wrong. This is why each part has its own padding.